### PR TITLE
Bug Fix - Correction for PR #9260 utf8mb4 / utf8 conversion for newly installed Joomla!

### PR DESCRIPTION
**Pull request by richard67**
#### Summary of Changes

This PR corrects a mistake I made with PR #9260 for the utf8mb4 / utf8 conversion to detect that it runs on a newly installed Joomla! and so nothing to be converted.

With PR #9260 I decided to initialize the conversion status in database table "#__utf8_conversion" with a value of 3 and later when running the conversion to adjust that to the expected value 1 or 2, depending on utf8mb4 supoort.

But if after a new installation nobody ever goes in the backend to "Extensions -> Manage -> Database -> Fix", and Joomla! is not updasted to a newer version, the status never is set to the final value.

This PR here corrects this by setting the correct conversion status at the end of a new installation.

So the changes to deal with the value 3 from PR #9260 can be undone with this PR here,. too.

This is pre-requisite for a later PR to correct the conversion in case of update and Database -> Fix in order not to run again when it is not necessary anymore by checking the conversion status in database (which without this PR here may still be 3 when migrating to another db server and so not correctly be handeled).
#### Testing Instructions

Hint: Replace "#__" by your db prefix.
1. Perform a new installation with latest staging, the source you can find here: [https://github.com/joomla/joomla-cms/archive/staging.zip](https://github.com/joomla/joomla-cms/archive/staging.zip)
2. Before login at backend, check in your database with e.g. phpMyAdmin the value of column "converted" in table "#__utf8_conversion". Result: Value = 3.
3. Login at backend and go to 'Extensions -> Manage -> Database'. Result: No database problems shown.
4. Press the "Fix" button even if no problems shown.
5. Check again in database the value of column "converted" in table "#__utf8_conversion". Result: Value = 1 if no utf8mb4 support (collations of core tables are uft8_unicode_ci) or 2 if utf8mb4 suport (collations of core tables are uft8mb4_unicode_ci).

Now repeat steps 1 to 3 (4 and 5 are not necessary now) with the branch for this patch.

Source for it is here: [https://github.com/richard67/joomla-cms/archive/correct-utf8mb4-new-install-2.zip](https://github.com/richard67/joomla-cms/archive/correct-utf8mb4-new-install-2.zip)

Result: The final value 1 or 2 of column "converted" in table "#__utf8_conversion" is already set in step 2, step 3 shows also in this test no database problems.

### CODE REVIEW

**Comment by richard67** [comment]
@wilsonge This PR here is a pre-requisite to solve your 3.5.0-blocker PR #9297 in the right way, and also to solve the other things which are to be done (with db fix only execute conversions once if not done yet, correct check of utf8mb4 conversion status by the input filter) in the right way, so this one here maybe needs the 3.5.0-blocker label, too?

The reason why I did not do all in this PR but will do the other stuff when this one here is done is that otherwise testing all in one would be too complicated.

So this one here has easy test from my point of view.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9303">issues.joomla.org/joomla-cms/9303</a>.</sub>

**Comment by richard67** [comment]
@andrepereiradasilva @infograf768 Could you test this one here if you find time? Would be important for 3.5.0, see my comment just before.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9303">issues.joomla.org/joomla-cms/9303</a>.</sub>

**Comment by andrepereiradasilva** [comment]
will do it today, just need to have some time.

**Comment by richard67** [comment]
Great, thanks in advance.

**Comment by Bodge-IT** [comment]
Hi Richard, thanks for the excellent intro & instructions. Really got me bought into the issue.
Unfortunately I got to step 2 and then couldn't see the __utf8_conversion table

**Comment by Bodge-IT** [comment]
My bad.  Table is there, this was down to me not paying attention to prefix and expecting new install to clear out old tables.

**Comment by anibalsanchez** [comment]
I have tested this item :white_check_mark: successfully on 636a66451232b604eecd7a00e215ffc612fe6037<br /><br />Test OK. Initial value 2.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9303">issues.joomla.org/joomla-cms/9303</a>.</sub>

**Comment by Bodge-IT** [comment]
I have tested this item :white_check_mark: successfully on 636a66451232b604eecd7a00e215ffc612fe6037<br /><br />Received Converted: 2 on utf8mb4 supported machine.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9303">issues.joomla.org/joomla-cms/9303</a>.</sub>

**Comment by richard67** [comment]
Thanks for testing.

joomla-brian-or-wilsonge-bot please RTC :smile: <hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9303">issues.joomla.org/joomla-cms/9303</a>.</sub>

**Comment by andrepereiradasilva** [comment]
hum i got the initial value of 3 in converted with clean install with your zip ...
```
Database Version: 5.5.5-10.0.24-MariaDB
Database Collation: utf8_general_ci
Database Connection Collation:  utf8mb4_general_ci
Tables Collation: utf8mb4_unicode_ci
```

**Comment by andrepereiradasilva** [comment]
ups worng zip file, sorry. will test agian

**Comment by andrepereiradasilva** [comment]
I have tested this item :white_check_mark: successfully on 636a66451232b604eecd7a00e215ffc612fe6037<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9303">issues.joomla.org/joomla-cms/9303</a>.</sub>

**Comment by richard67** [comment]
@andrepereiradasilva Thanks for testing.

Just a hint: Collations should be utf8mb4_**unicode**_ci for the core tables of a new install with my zip.

**Comment by andrepereiradasilva** [comment]
sure and they are.

BTW, not related to this PR, but shouldn't the other joomla core extensions use utf8mb4_unicode_ci too? 

That i know of, that use database tables there are only:
- Patchtester Component: https://github.com/joomla-extensions/patchtester<br/>(already created in utf8mb4_unicode_ci)

- Weblinks Component: https://github.com/joomla-extensions/weblinks<br/>(not yet in utf8mb4_unicode_ci)

**Comment by richard67** [comment]
@andrepereiradasilva Yes, they should. They should use **utf8mb4_unicode_ci** (which will be changed to utf8_unicode_ci in sql statements by the database driver in case of no utf8mb4 support).

I had seen you had prepared some PRs for some of the extensions a while ago with _general_. Those would have to be changed, yes.

Btw. thanks for your connection collation info above, just am preparing a new PR for the joomla database driver to set connection collation to utf8_unicode_ci or utf8mb4_unicode_ci, too (not _general_ like now). Was not aware of that. The _unicode_ sorting is always correct while the _general_ is a compromise, and I think we need it in the connection settings too, to get things sorted right in russian and so on. But I will ask Eli in my new PR later to confirm that.

**Comment by andrepereiradasilva** [comment]
> I had seen you had prepared some PRs for some of the extensions a while ago with general. Those would have to be changed, yes.

i only did for the patchtester and with utf8mb4_unicode_ci
See https://github.com/joomla-extensions/patchtester/pull/127/files

**Comment by richard67** [comment]
So patchtester is OK, great.

**Comment by andrepereiradasilva** [comment]
yeah, as soom as mbabker releases a beta (or stable) they will be CREATED (only clean install) with utf8mb4_unicode_ci. only weblinks missing i think.

**Comment by richard67** [comment]
Yes, conversion of existing extensions tables is an open point still.

**Comment by andrepereiradasilva** [comment]
not for the patchtester not, it has no upgrade path yet. so no need.

**Comment by richard67** [comment]
I see. Forget about my new PR I wanna make for setting the connection collation also to unicode sorting. This seems to be a bit complicated, see discussion on [https://mathiasbynens.be/notes/mysql-utf8mb4](https://mathiasbynens.be/notes/mysql-utf8mb4)

**Comment by andrepereiradasilva** [comment]
ok, so maintainers please RTC this and let's move on :)

File: installation/model/database.php [hunk 1]
```diff
@@ -517,6 +517,29 @@ public function createTables($options)
 			return false;
 		}
 
+		// Get query object for later database access
+		$query = $db->getQuery(true);
+
+		// MySQL only: Attempt to update the table #__utf8_conversion.
+		if (($type == 'mysql') || ($type == 'mysqli') || ($type == 'pdomysql'))
```
Comment (wilsonge) [inline installation/model/database.php]: Can we be consistent and use the same server type check as here please? :) https://github.com/joomla/joomla-cms/blob/staging/administrator/components/com_installer/models/database.php#L270-L275

**Comment by wilsonge** [comment]
Setting RTC - but can we just fix up my comment richard - as long as that's the only change won't need redoing of the tests

**Comment by richard67** [comment]
@wilsonge I uses the same check as already existent in this file a few lined below mine. Was not sure if the other one is supported. Question: Shall I change all of them? Or only the one I've added?

**Comment by richard67** [comment]
@wilsonge sorry, meant above, not below. I took it from line 493 where is decided which joomla.sql to use. Please advise if I shall change all or only mine or none.

**Comment by wilsonge** [comment]
The method used in the database fixer was new for 3.5 - so the other one below just never got updated. Can we please just update this one instance so UTF8MB4 is consistent - we can do the other one in a new PR :)

**Comment by richard67** [comment]
@wilsonge Done.

**Comment by 810** [comment]
when i do a new install with default utf8mb4 collation, then after the install has changed to utf8_general_ci as default collation.

**Comment by richard67** [comment]
@810 Sorry, do not understand. You mean on a database not supporting utf8mb4? And sure it is general, not unicode in the middle ? And do you mean database default? This we not change, we change only for all core tables individually.

**Comment by 810** [comment]
i will explain what i mean.

1) create new database with  utf8mb4_unicode_ci collation
2) Run installer (database mysqli)
3) After install the default collation is set to utf8_general_ci see image
ee

**Comment by richard67** [comment]
@810 I see. Well, is not related to this PR or our utf8mb4 conversions, because those definitely not touch thd database default collation, and it does not do any harm for Joomla! because Jomla! and the official extension(s) specify collation for every table on table creation.

So it is not related to this PR here, but feel free to make a new issue for that in the issue tracker, or a new PR.

I'll put on my to do list to check if I find somewhere in Joomla! if this database default collation is set somwhere (php code or sql scripts), but if you are faster and find it, let me know.

**Comment by andrepereiradasilva** [comment]
I know is not related to this PR, but when you install joomla without a database pre created, 

In 
https://github.com/joomla/joomla-cms/blob/staging/libraries/joomla/database/driver.php#L858

```
CREATE DATABASE dabame CHARACTER SET `utf8mb4` COLLATE utf8mb4_unicode_ci;
```

should be something like this for mysql i think

**Comment by 810** [comment]
i think its line 766 let me check it

**Comment by richard67** [comment]
The driver.php contains "_general" in several code and comment lines, code lines are 737 and 766. Should be changed all I think but don't ask me for testing instructions. Who wants to do a PR for changing the driver.php?

**Comment by 810** [comment]
i have changed them,but still not working. but for this one i can do a pr

**Comment by andrepereiradasilva** [comment]
i think is in the line i said before, because is when the database is created, not the tables.

**Comment by 810** [comment]
https://github.com/joomla/joomla-cms/pull/9311/files

@andrepereiradasilva i have added it on your line, but is not working, so there is somewhere else the issue

**Pull request merged**